### PR TITLE
Fix the race condition that causes syncer missing in the syncerMap

### PR DESCRIPTION
**Pull request by ruixiansong**
/assign @swetharepakula 
/cc @gauravkghildiyal

### CODE REVIEW

File: pkg/neg/manager_test.go [hunk 1]
```diff
@@ -330,6 +330,43 @@ func TestEnsureAndStopSyncer(t *testing.T) {
...
+	portInfo := types.PortInfo{PortTuple: negtypes.SvcPortTuple{Name: portName, Port: svcPort, TargetPort: targetPort}, NegName: negName}
+
+	portMap[negtypes.PortInfoMapKey{ServicePort: svcPort}] = portInfo
+
+	manager.serviceLister.Add(&v1.Service{ObjectMeta: metav1.ObjectMeta{Namespace: namespace, Name: name}})
+
+	svcKey := serviceKey{namespace: namespace, name: name}
+	negSyncerKey := manager.getSyncerKey(namespace, name, negtypes.PortInfoMapKey{ServicePort: svcPort}, portInfo)
+	manager.svcPortMap[svcKey] = portMap
+	manager.svcPortMap = map[serviceKey]negtypes.PortInfoMap{
```
Comment (sawsa307) [inline pkg/neg/manager_test.go:352]: I suppose this line is doing the same thing as the previous line?

**Review by sawsa307 — COMMENTED** [review]
One nit, everything else looks good to me. Will leave final lgtm to @swetharepakula and @gauravkghildiyal 

/approve

File: pkg/neg/manager_test.go (hunk 1 above)
Comment (ruixiansong) [reply pkg/neg/manager_test.go]: Thanks! Done!

**Comment by gauravkghildiyal** [comment]
Thanks @ruixiansong 

It's possible that I am misunderstanding something, but to me it seems that this still has the same race condition:

- The first time around when the syncer fails to start, because we've added it to the syncerMap already, the second Ensure() corresponding to the `samePorts` will never get executed.

Assuming that we still have the race, we likely also need to improve our unit test to detect the problem?

**Comment by ruixiansong** [comment]
> Thanks @ruixiansong
> 
> It's possible that I am misunderstanding something, but to me it seems that this still has the same race condition:
> 
> * The first time around when the syncer fails to start, because we've added it to the syncerMap already, the second Ensure() corresponding to the `samePorts` will never get executed.
> 
> Assuming that we still have the race, we likely also need to improve our unit test to detect the problem?

Thanks @gauravkghildiyal for reviewing this.

I'm assuming the syncer in the map will finally got GCed. Then it won't be in the syncerMap and then we can recreate it here.

Or we need to start the syncer here again before it got GCed?

**Comment by gauravkghildiyal** [comment]
> > Thanks @ruixiansong
> > It's possible that I am misunderstanding something, but to me it seems that this still has the same race condition:
> > 
> > * The first time around when the syncer fails to start, because we've added it to the syncerMap already, the second Ensure() corresponding to the `samePorts` will never get executed.
> > 
> > Assuming that we still have the race, we likely also need to improve our unit test to detect the problem?
> 
> Thanks @gauravkghildiyal for reviewing this.
> 
> I'm assuming the syncer in the map will finally got GCed. Then it won't be in the syncerMap and then we can recreate it here.
> 
> Or we need to start the syncer here again before it got GCed?

Thanks for clarification. Yes that does seem to resolve the race.

Though doesn't the wait for the GC introduce the following degrading behaviour:

1. Service is created. Syncers are running well and healthy
2. Service is deleted. We allow sufficient time for the Service to be deleted (maybe like 1 minute)
3. After that 1 minute, we recreate the Service.
4. It's possible that the GC hasn't happened by this point which means the syncerMap still has the previous syncer. 
5. When we do EnsureSyncer() again, because syncerMap still has the previous syncer (which isn't GC'd yet), we never start it again. In addition, we also don't return any errors.
6. This means that for the recreated Service, it likely has to wait for the next periodic resync of the Service to properly start the syncers...which means a delay of 10 minutes at minimum.

Sorry for making this painful, but the changes here touch a critical lifecycle, and I want to make sure we have all bases covered. (So I may end up calling out problems which I may not have fully vetted)

File: pkg/neg/manager.go [hunk 2]
```diff
@@ -199,7 +199,19 @@ func (manager *syncerManager) EnsureSyncers(namespace, name string, newPorts neg
...
+			if err := manager.EnsureSyncer(namespace, name, svcPort, portInfo); err != nil {
+				errList = append(errList, err)
+				errorSyncers += 1
+			}
+			successfulSyncers += 1
+			continue
+		}
 		// To reduce the possibility of NEGs being leaked, ensure a SvcNeg CR exists for every
 		// desired port.
 		if err := manager.ensureSvcNegCR(key, portInfo); err != nil {
```
Comment (gauravkghildiyal) [inline pkg/neg/manager.go:217]: We're doing an ensureSvcNEGCR here, and we're also doing one within EnsureSyncer().

Doesn't this mean that we're doing the exact same thing for both `samePorts` and `addPorts`? Assuming `newPorts = samePorts + addPorts`, maybe we can refactor this to just process `newPorts`? (Let's double check)

**Comment by ruixiansong** [comment]
> > > Thanks @ruixiansong
> > > It's possible that I am misunderstanding something, but to me it seems that this still has the same race condition:
> > > 
> > > * The first time around when the syncer fails to start, because we've added it to the syncerMap already, the second Ensure() corresponding to the `samePorts` will never get executed.
> > > 
> > > Assuming that we still have the race, we likely also need to improve our unit test to detect the problem?
> > 
> > 
> > Thanks @gauravkghildiyal for reviewing this.
> > I'm assuming the syncer in the map will finally got GCed. Then it won't be in the syncerMap and then we can recreate it here.
> > Or we need to start the syncer here again before it got GCed?
> 
> Thanks for clarification. Yes that does seem to resolve the race.
> 
> Though doesn't the wait for the GC introduce the following degrading behaviour:
> 
> 1. Service is created. Syncers are running well and healthy
> 2. Service is deleted. We allow sufficient time for the Service to be deleted (maybe like 1 minute)
> 3. After that 1 minute, we recreate the Service.
> 4. It's possible that the GC hasn't happened by this point which means the syncerMap still has the previous syncer.
> 5. When we do EnsureSyncer() again, because syncerMap still has the previous syncer (which isn't GC'd yet), we never start it again. In addition, we also don't return any errors.
> 6. This means that for the recreated Service, it likely has to wait for the next periodic resync of the Service to properly start the syncers...which means a delay of 10 minutes at minimum.
> 
> Sorry for making this painful, but the changes here touch a critical lifecycle, and I want to make sure we have all bases covered. (So I may end up calling out problems which I may not have fully vetted)

That makes sense, updated the EnsureSyncers() function to restart the stopped syncers to avoid the degrading behaviour.

File: pkg/neg/manager.go (hunk 2 above)
Comment (ruixiansong) [reply pkg/neg/manager.go:208]: Yes that makes sense, I completely refactored this one and combined those with one loop for all newPorts.

**Comment by gauravkghildiyal** [comment]
Thanks!

/lgtm

**Pull request merged**